Session.cc

```diff
@@ -744,33 +757,36 @@ void Session::sendCubeHistogramData(const CARTA::SetHistogramRequirements& messa
                             createCubeHistogramMessage(histogramProgressMsg, fileId, stokes, progress);
+                            CARTA::Histogram* histogram = histogramMessage.add_histograms();  // blank
                             sendFileEvent(fileId, "REGION_HISTOGRAM_DATA", requestId, histogramProgressMsg);
                             tStart = tEnd;
                         }
                     }
+                    // save min,max in cube region
+                    if (histogramProgress > HISTOGRAM_CANCEL)
+                        frames.at(fileId)->setRegionMinMax(regionId, channel, stokes, cubemin, cubemax);
+
+                    // check cancel and proceed
                     if (histogramProgress > HISTOGRAM_CANCEL) {
-                        // send message in case min/max took < 2s; progress is half done
+                        // send progress message: half done
                         float progress = 0.50;
-                        CARTA::RegionHistogramData halfProgressMsg;
-                        createCubeHistogramMessage(halfProgressMsg, fileId, stokes, progress);
-                        sendFileEvent(fileId, "REGION_HISTOGRAM_DATA", requestId, halfProgressMsg);
-
-                        // get histogram bins for each channel
-                        std::valarray<int> cubeBins(0, numbins);  // accumulate bin counts
-                        // create message for each channel histogram (not sent)
-                        CARTA::RegionHistogramData chanHistogramMessage;
-                        createCubeHistogramMessage(chanHistogramMessage, fileId, stokes, progress);
-                        for (size_t chan=0; chan < nchan; ++chan) {
-                            // get CARTA::Histogram for this channel
-                            chanHistogramMessage.clear_histograms();
-                            std::vector<float> data = frames.at(fileId)->getImageChanData(chan);
-                            frames.at(fileId)->fillChannelHistogramData(chanHistogramMessage, data, chan, numbins, minval, maxval);
-                            auto chanHistogram = chanHistogramMessage.histograms(0);
+                        histogramMessage.set_progress(progress);
+                        sendFileEvent(fileId, "REGION_HISTOGRAM_DATA", requestId, histogramMessage);
 
+                        // get histogram bins for each channel and accumulate bin counts
+                        std::vector<int> cubeBins;
+                        CARTA::Histogram chanHistogram;  // histogram for each channel
+                        for (size_t chan=0; chan < nchan; ++chan) { 
+                            frames.at(fileId)->calcRegionHistogram(regionId, chan, stokes, numbins, cubemin,
+                                cubemax, chanHistogram);
                             // add channel bins to cube bins
                             std::vector<int> channelBins = {chanHistogram.bins().begin(), chanHistogram.bins().end()};
```

Comment:
Is this faster than accessing `chanHistogram.bins()` directly below? We should perhaps profile both options to compare them.

Reply:
I will look into profiling this; I think it's OK the way it is for now.

Reply:
Will look into this.

Reply:
I've dug into the generated protobuf code for getting  and setting  elements of  a repeated field and it's basically inlining an array access, so I'd imagine using it directly should be the fastest method.

Reply:
Changed to using iterators for addition, tested old/new on 2 datasets [1920,1920,478] and [4224,1824,222].  Using transform() reduced the total time to calculate cube histogram bins by ~250 ms (total time of ~104 s and ~112 s).

std::transform(chanHistogram.bins().begin(), chanHistogram.bins().end(), cubeBins.begin(),
                                    cubeBins.begin(), std::plus<int>());